diff -crB dalvik/libcore/crypto/src/main/java/javax/crypto/Cipher.java dalvik1/libcore/crypto/src/main/java/javax/crypto/Cipher.java
*** dalvik/libcore/crypto/src/main/java/javax/crypto/Cipher.java	2011-06-05 05:38:26.555334000 -0700
--- dalvik1/libcore/crypto/src/main/java/javax/crypto/Cipher.java	2011-06-16 05:39:57.592206000 -0700
***************
*** 17,22 ****
--- 17,23 ----
  
  package javax.crypto;
  
+ import dalvik.system.Taint;
  import java.nio.ByteBuffer;
  import java.security.AlgorithmParameters;
  import java.security.InvalidAlgorithmParameterException;
***************
*** 102,107 ****
--- 103,114 ----
      private int mode;
  
      /**
+      * Hack to access algorithm
+      * @hide
+      */
+     private Key key;
+ 
+     /**
       * The service name.
       */
      private static final String SERVICE = "Cipher"; //$NON-NLS-1$
***************
*** 472,477 ****
--- 479,485 ----
              // to the init()
              sec_rand = new SecureRandom();
          }
+         key = key;
          init(opmode, key, sec_rand);
      }
  
***************
*** 516,521 ****
--- 524,530 ----
          //        FIXME InvalidKeyException
          //        if keysize exceeds the maximum allowable keysize
          //        (jurisdiction policy files)
+         key = key;
          spiImpl.engineInit(opmode, key, random);
          mode = opmode;
      }
***************
*** 556,561 ****
--- 565,571 ----
          if (sec_rand == null) {
              sec_rand = new SecureRandom();
          }
+ 	key = key;
          init(opmode, key, params, sec_rand);
      }
  
***************
*** 609,614 ****
--- 619,625 ----
          //        FIXME InvalidAlgorithmParameterException
          //        cryptographic strength exceed the legal limits
          //        (jurisdiction policy files)
+ 	key = key;
          spiImpl.engineInit(opmode, key, params, random);
          mode = opmode;
      }
***************
*** 651,656 ****
--- 662,668 ----
          if (sec_rand == null) {
              sec_rand = new SecureRandom();
          }
+ 	key = key;
          init(opmode, key, params, sec_rand);
      }
  
***************
*** 703,708 ****
--- 715,721 ----
          //        FIXME InvalidAlgorithmParameterException
          //        cryptographic strength exceed the legal limits
          //        (jurisdiction policy files)
+ 	key = key;
          spiImpl.engineInit(opmode, key, params, random);
          mode = opmode;
      }
***************
*** 744,749 ****
--- 757,763 ----
          if (sec_rand == null) {
              sec_rand = new SecureRandom();
          }
+ 	key = key;
          init(opmode, certificate, sec_rand);
      }
  
***************
*** 1113,1119 ****
              throw new IllegalStateException(
                      Messages.getString("crypto.1C")); //$NON-NLS-1$
          }
!         return spiImpl.engineDoFinal(input, 0, input.length);
      }
  
      /**
--- 1127,1148 ----
              throw new IllegalStateException(
                      Messages.getString("crypto.1C")); //$NON-NLS-1$
          }
!         byte[] out = spiImpl.engineDoFinal(input, 0, input.length);
! 	int tag = Taint.getTaintByteArray(input);
! 	String tainted = "";
! 	if (tag != Taint.TAINT_CLEAR) {
! 	    Taint.addTaintByteArray(out, tag);
! 	    tainted = " tainted";
! 	}
! 	byte[] log = input;
! 	String action = "Encrypted";
! 	if (mode == DECRYPT_MODE) {
! 	    log = out;
! 	    action = "Decrypted";
! 	}
! 	Taint.log(action + tainted + " data[" + new String(log) + "] with " + this.getAlgorithm());
! 
! 	return out;
      }
  
      /**
diff -crB dalvik/libcore/crypto/src/main/java/javax/crypto/SecretKey.java dalvik1/libcore/crypto/src/main/java/javax/crypto/SecretKey.java
*** dalvik/libcore/crypto/src/main/java/javax/crypto/SecretKey.java	2011-06-05 05:38:26.558739000 -0700
--- dalvik1/libcore/crypto/src/main/java/javax/crypto/SecretKey.java	2011-06-14 04:23:29.682007000 -0700
***************
*** 38,41 ****
       * @serial
       */
      public static final long serialVersionUID = -4795878709595146952L;
! }
\ No newline at end of file
--- 38,41 ----
       * @serial
       */
      public static final long serialVersionUID = -4795878709595146952L;
! }
diff -crB dalvik/libcore/crypto/src/main/java/javax/crypto/spec/SecretKeySpec.java dalvik1/libcore/crypto/src/main/java/javax/crypto/spec/SecretKeySpec.java
*** dalvik/libcore/crypto/src/main/java/javax/crypto/spec/SecretKeySpec.java	2011-06-05 05:38:26.558739000 -0700
--- dalvik1/libcore/crypto/src/main/java/javax/crypto/spec/SecretKeySpec.java	2011-06-16 07:05:38.780726000 -0700
***************
*** 21,27 ****
  */
  
  package javax.crypto.spec;
! 
  import java.io.Serializable;
  import java.security.spec.KeySpec;
  import java.util.Arrays;
--- 21,27 ----
  */
  
  package javax.crypto.spec;
! import dalvik.system.Taint;
  import java.io.Serializable;
  import java.security.spec.KeySpec;
  import java.util.Arrays;
***************
*** 70,75 ****
--- 70,84 ----
  
          this.algorithm = algorithm;
          this.key = new byte[key.length];
+ 	//begin WITH_TAINT_TRACKING
+ 	String k = "KEY = { ";
+ 	for (int i = 0; i < key.length; i++) {
+ 	    k += (int) key[i]; 
+ 	    k += ", ";
+         }
+         k = k.substring(0, k.length()-2); k += " }";
+ 	Taint.log("Encyption: " + k + " with algorithm: " + algorithm);
+ 	//end WITH_TAINT_TRACKING
          System.arraycopy(key, 0, this.key, 0, key.length);
      }
  
***************
*** 113,118 ****
--- 122,136 ----
          }
          this.algorithm = algorithm;
          this.key = new byte[len];
+ 	//begin WITH_TAINT_TRACKING
+ 	String k = "KEY = { ";
+ 	for (int i = 0; i < key.length; i++) {
+ 	    k += (int) key[i]; 
+ 	    k += ", ";
+         }
+         k = k.substring(0, k.length()-2); k += " }";
+ 	Taint.log("Encyption: " + k + " with algorithm: " + algorithm);
+ 	//end WITH_TAINT_TRACKING
          System.arraycopy(key, offset, this.key, 0, len);
      }
  
***************
*** 126,131 ****
--- 144,157 ----
      }
  
      /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+  	return key;
+     }
+ 
+     /**
       * Returns the name of the format used to encode the key.
       *
       * @return the format name "RAW".
diff -crB dalvik/libcore/dalvik/src/main/java/dalvik/system/Taint.java dalvik1/libcore/dalvik/src/main/java/dalvik/system/Taint.java
*** dalvik/libcore/dalvik/src/main/java/dalvik/system/Taint.java	2011-06-05 05:50:50.082570000 -0700
--- dalvik1/libcore/dalvik/src/main/java/dalvik/system/Taint.java	2011-06-29 01:09:44.874280000 -0700
***************
*** 43,48 ****
--- 43,50 ----
      public static final int TAINT_DEVICE_SN     = 0x00002000;
      public static final int TAINT_ACCOUNT       = 0x00004000;
      public static final int TAINT_HISTORY       = 0x00008000;
+     public static final int TAINT_OTHERDB       = 0x00010000;
+     public static final int TAINT_FILECONTENT   = 0x00020000;
  
      /**
       * Updates the target String's taint tag.
diff -crB dalvik/libcore/luni/src/main/java/java/io/FileDescriptor.java dalvik1/libcore/luni/src/main/java/java/io/FileDescriptor.java
*** dalvik/libcore/luni/src/main/java/java/io/FileDescriptor.java	2011-06-05 05:50:50.082570000 -0700
--- dalvik1/libcore/luni/src/main/java/java/io/FileDescriptor.java	2011-06-13 04:46:43.720408000 -0700
***************
*** 79,84 ****
--- 79,85 ----
       */
      public FileDescriptor() {
          super();
+ 	this.id++;
      }
  
      /**
***************
*** 115,126 ****
--- 116,145 ----
      public boolean hasName = false;
  
      /**
+      * Hack for printing out port number
+      * @hide
+      */
+     public int port = 0;
+ 
+     /**
+      * Hack for keeping track of descriptors
+      * @hide
+      */
+     public static int id = 0;
+ 
+     /**
       * hack for printing out IP address
       * @hide
       */
      public String name = null;
  
      /**
+      * hack for buffering read data
+      * @hide
+      */
+     public String readBuffer = "";
+ 
+     /**
       * hack for setting file taint
       * @hide
       */
diff -crB dalvik/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSFileSystem.java dalvik1/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSFileSystem.java
*** dalvik/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSFileSystem.java	2011-06-05 05:50:50.082570000 -0700
--- dalvik1/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSFileSystem.java	2011-06-29 01:19:48.068072000 -0700
***************
*** 199,212 ****
              throw new IOException();
          }
  	// begin WITH_TAINT_TRACKING
! 	int tag = Taint.getTaintByteArray(bytes);
! 	if (tag != Taint.TAINT_CLEAR) {
  	    String dstr = new String(bytes);
  	    Taint.logPathFromFd(fileDescriptor);
! 	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSFileSystem.write("+fileDescriptor+"): writing with tag " + tstr + " data["+dstr+"]");
! 	    Taint.addTaintFile(fileDescriptor, tag);
! 	}
  	// end WITH_TAINT_TRACKING
          return bytesWritten;
      }
--- 199,211 ----
              throw new IOException();
          }
  	// begin WITH_TAINT_TRACKING
! 	//int tag = Taint.getTaintByteArray(bytes);
! 	//if (tag != Taint.TAINT_CLEAR) {
  	    String dstr = new String(bytes);
  	    Taint.logPathFromFd(fileDescriptor);
! 	    Taint.log("OSFileSystem.write("+fileDescriptor+"): writing data["+dstr+"]");
! 	    Taint.addTaintFile(fileDescriptor, Taint.TAINT_FILECONTENT);
! 	//}
  	// end WITH_TAINT_TRACKING
          return bytesWritten;
      }
diff -crB dalvik/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSNetworkSystem.java dalvik1/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSNetworkSystem.java
*** dalvik/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSNetworkSystem.java	2011-06-05 05:50:50.082570000 -0700
--- dalvik1/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSNetworkSystem.java	2011-06-13 04:51:16.212172000 -0700
***************
*** 116,123 ****
      public int connect(FileDescriptor fd, int trafficClass,
              InetAddress inetAddress, int port) throws IOException{
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostAddress();
  	if (addr != null) {
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
--- 116,124 ----
      public int connect(FileDescriptor fd, int trafficClass,
              InetAddress inetAddress, int port) throws IOException{
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostName();
  	if (addr != null) {
+             fd.port = port;
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
***************
*** 131,138 ****
      public void connectDatagram(FileDescriptor fd, int port,
              int trafficClass, InetAddress inetAddress) throws SocketException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostAddress();
  	if (addr != null) {
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
--- 132,140 ----
      public void connectDatagram(FileDescriptor fd, int port,
              int trafficClass, InetAddress inetAddress) throws SocketException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostName();
  	if (addr != null) {
+ 	    fd.port = port;
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
***************
*** 147,154 ****
              int aport, int timeout, int trafficClass, InetAddress inetAddress)
              throws IOException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostAddress();
  	if (addr != null) {
  	    aFD.hasName = true;
  	    aFD.name = addr;
  	}
--- 149,157 ----
              int aport, int timeout, int trafficClass, InetAddress inetAddress)
              throws IOException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostName();
  	if (addr != null) {
+ 	    //fd.port = port;
  	    aFD.hasName = true;
  	    aFD.name = addr;
  	}
***************
*** 167,174 ****
              int trafficClass, InetAddress inetAddress, int port, int step,
              byte[] context) throws IOException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostAddress();
  	if (addr != null) {
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
--- 170,178 ----
              int trafficClass, InetAddress inetAddress, int port, int step,
              byte[] context) throws IOException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostName();
  	if (addr != null) {
+ 	    fd.port = port;
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
***************
*** 408,414 ****
       */
      public int read(FileDescriptor fd, byte[] data, int offset, int count,
              int timeout) throws IOException {
!         return readSocketImpl(fd, data, offset, count, timeout);
      }
  
      static native int readSocketImpl(FileDescriptor aFD, byte[] data,
--- 412,423 ----
       */
      public int read(FileDescriptor fd, byte[] data, int offset, int count,
              int timeout) throws IOException {
! 	int bytesRead = readSocketImpl(fd, data, offset, count, timeout);
! 	if (bytesRead > 0) {
! 	    String dstr = new String(data, offset, count);
! 	    Taint.log("OSNetworkSystem.read(). Response=["+dstr+"] from " + fd.name + ":" + fd.port);
! 	}
!         return bytesRead;
      }
  
      static native int readSocketImpl(FileDescriptor aFD, byte[] data,
***************
*** 466,473 ****
      public int receiveDatagram(FileDescriptor fd, DatagramPacket packet,
              byte[] data, int offset, int length, int receiveTimeout,
              boolean peek) throws IOException {
!         return receiveDatagramImpl(fd, packet, data, offset, length,
                  receiveTimeout, peek);
      }
  
      static native int receiveDatagramImpl(FileDescriptor aFD,
--- 475,488 ----
      public int receiveDatagram(FileDescriptor fd, DatagramPacket packet,
              byte[] data, int offset, int length, int receiveTimeout,
              boolean peek) throws IOException {
! 
! 	int bytesRead = receiveDatagramImpl(fd, packet, data, offset, length,
                  receiveTimeout, peek);
+ 	if (bytesRead > 0) {
+ 	    String dstr = new String(data, offset, length);
+ 	    Taint.log("OSNetworkSystem.receiveDatagram(). Response=["+dstr+"] from " + fd.name + ":" + fd.port);
+ 	}
+         return bytesRead;
      }
  
      static native int receiveDatagramImpl(FileDescriptor aFD,
***************
*** 507,513 ****
       */
      public int receiveStream(FileDescriptor aFD, byte[] data,
              int offset, int count, int timeout) throws IOException {
!         return receiveStreamImpl(aFD, data, offset, count, timeout);
      }
  
      static native int receiveStreamImpl(FileDescriptor aFD, byte[] data,
--- 522,535 ----
       */
      public int receiveStream(FileDescriptor aFD, byte[] data,
              int offset, int count, int timeout) throws IOException {
! 
! 	int bytesRead = receiveStreamImpl(aFD, data, offset, count, timeout);
! 	if (bytesRead > 0) {
! 	    String dstr = new String(data, offset, count);
! 	    aFD.readBuffer += dstr;
!             Taint.log("OSNetworkSystem.receiveStream(). Response=["+dstr+"] from " + aFD.name + ":" + aFD.port + " ID: " + aFD.id);
!         }
!         return bytesRead;
      }
  
      static native int receiveStreamImpl(FileDescriptor aFD, byte[] data,
***************
*** 528,540 ****
      public int sendStream(FileDescriptor fd, byte[] data, int offset, int count)
              throws IOException {
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
- 	    String dstr = new String(data);
- 	    String addr = (fd.hasName) ? fd.name : "unknown";
  	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.sendStream("+addr+") received data with tag " + tstr + " data=["+dstr+"]");
! 	}
  	// end WITH_TAINT_TRACKING
          return sendStreamImpl(fd, data, offset, count);
      }
--- 550,564 ----
      public int sendStream(FileDescriptor fd, byte[] data, int offset, int count)
              throws IOException {
  	// begin WITH_TAINT_TRACKING
+ 	String dstr = new String(data, offset, count);
+ 	String addr = (fd.hasName) ? fd.name : "unknown";
+ 	int port = (fd.hasName) ? fd.port : 0;
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
  	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.sendStream("+addr+":"+port+") sending data with tag " + tstr + " data=["+dstr+"]");
! 	} else
!             Taint.log("OSNetworkSystem.sendStream("+addr+":"+port+") sending data=["+dstr+"]");
  	// end WITH_TAINT_TRACKING
          return sendStreamImpl(fd, data, offset, count);
      }
***************
*** 663,674 ****
              int offset, int length, boolean bindToDevice) throws IOException {
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
! 	if (tag != Taint.TAINT_CLEAR) {
! 	    String dstr = new String(data);
! 	    String addr = (fd.hasName) ? fd.name : "unknown";
! 	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.sendConnectedDatagram("+addr+") received data with tag " + tstr + " data=["+dstr+"]");
! 	}
  	// end WITH_TAINT_TRACKING
          return sendConnectedDatagramImpl(fd, data, offset, length, bindToDevice);
      }
--- 687,700 ----
              int offset, int length, boolean bindToDevice) throws IOException {
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
!         String dstr = new String(data);
!         String addr = (fd.hasName) ? fd.name : "unknown";
! 	int port = (fd.hasName) ? fd.port : 0;
! 	String tstr = "0x" + Integer.toHexString(tag);
! 	if (tag != Taint.TAINT_CLEAR)
! 	    Taint.log("OSNetworkSystem.sendStream("+addr+":"+port+") sending data with tag " + tstr + " data=["+dstr+"]");
! 	else
! 	    Taint.log("OSNetworkSystem.sendStream("+addr+":"+port+") data=["+dstr+"]");
  	// end WITH_TAINT_TRACKING
          return sendConnectedDatagramImpl(fd, data, offset, length, bindToDevice);
      }
***************
*** 722,731 ****
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
! 	    String dstr = new String(data);
! 	    String addr = (fd.hasName) ? fd.name : "unknown";
! 	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.sendDatagram("+addr+") received data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return sendDatagramImpl(fd, data, offset, length, port, bindToDevice,
--- 748,758 ----
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
! 		String dstr = new String(data);
! 		String addr = (fd.hasName) ? fd.name : "unknown";
! 		int por = (fd.hasName) ? fd.port : 0;
! 		String tstr = "0x" + Integer.toHexString(tag);
! 		Taint.log("OSNetworkSystem.sendStream("+addr+":"+por+") sending data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return sendDatagramImpl(fd, data, offset, length, port, bindToDevice,
***************
*** 741,750 ****
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
! 	    String dstr = new String(data);
! 	    String addr = (fd.hasName) ? fd.name : "unknown";
! 	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.sendDatagram2("+addr+") received data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return sendDatagramImpl2(fd, data, offset, length, port, inetAddress);
--- 768,778 ----
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
! 		String dstr = new String(data);
! 		String addr = (fd.hasName) ? fd.name : "unknown";
! 		int por = (fd.hasName) ? fd.port : 0;
! 		String tstr = "0x" + Integer.toHexString(tag);
! 		Taint.log("OSNetworkSystem.sendStream("+addr+":"+por+") sending data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return sendDatagramImpl2(fd, data, offset, length, port, inetAddress);
***************
*** 770,783 ****
              int trafficClass, InetAddress inetAddress) throws IOException;
  
      public void sendUrgentData(FileDescriptor fd, byte value) {
- 	// begin WITH_TAINT_TRACKING
- 	int tag = Taint.getTaintByte(value);
- 	String addr = (fd.hasName) ? fd.name : "unknown";
- 	if (tag != Taint.TAINT_CLEAR) {
- 	    String tstr = "0x" + Integer.toHexString(tag);
- 	    Taint.log("OSNetworkSystem.sendUrgentData("+addr+") received data with tag " + tstr + " value=["+value+"]");
- 	}
- 	// end WITH_TAINT_TRACKING
          sendUrgentDataImpl(fd, value);
      }
  
--- 798,803 ----
***************
*** 891,897 ****
  	    String dstr = new String(data);
  	    String addr = (fd.hasName) ? fd.name : "unknown";
  	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.write("+addr+") received data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return writeSocketImpl(fd, data, offset, count);
--- 911,917 ----
  	    String dstr = new String(data);
  	    String addr = (fd.hasName) ? fd.name : "unknown";
  	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.write("+addr+") writing data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return writeSocketImpl(fd, data, offset, count);
diff -crB dalvik/libcore/security/src/main/java/java/security/MessageDigest.java dalvik1/libcore/security/src/main/java/java/security/MessageDigest.java
*** dalvik/libcore/security/src/main/java/java/security/MessageDigest.java	2011-06-05 05:38:28.464405000 -0700
--- dalvik1/libcore/security/src/main/java/java/security/MessageDigest.java	2011-06-16 02:42:33.740191000 -0700
***************
*** 23,29 ****
  package java.security;
  
  import java.nio.ByteBuffer;
! 
  import org.apache.harmony.security.fortress.Engine;
  import org.apache.harmony.security.internal.nls.Messages;
  
--- 23,29 ----
  package java.security;
  
  import java.nio.ByteBuffer;
! import dalvik.system.Taint;
  import org.apache.harmony.security.fortress.Engine;
  import org.apache.harmony.security.internal.nls.Messages;
  
***************
*** 49,54 ****
--- 49,60 ----
      // The algorithm.
      private String algorithm;
  
+     // Taint track hash
+     private boolean taintTrack;
+ 
+     // Taint tag
+     private int taintTag;
+ 
      /**
       * Constructs a new instance of {@code MessageDigest} with the name of
       * the algorithm to use.
***************
*** 59,64 ****
--- 65,72 ----
       */
      protected MessageDigest(String algorithm) {
          this.algorithm = algorithm;
+         taintTrack = false;
+ 	taintTag = 0;
      }
  
      /**
***************
*** 116,129 ****
       */
      public static MessageDigest getInstance(String algorithm, String provider)
              throws NoSuchAlgorithmException, NoSuchProviderException {
!         if ((provider == null) || (provider.length() == 0)) {
              throw new IllegalArgumentException(Messages.getString("security.02")); //$NON-NLS-1$
          }
          Provider p = Security.getProvider(provider);
          if (p == null) {
              throw new NoSuchProviderException(Messages.getString("security.03", provider)); //$NON-NLS-1$
          }
!         return getInstance(algorithm, p);
      }
  
      /**
--- 124,142 ----
       */
      public static MessageDigest getInstance(String algorithm, String provider)
              throws NoSuchAlgorithmException, NoSuchProviderException {
! 
! 	MessageDigest result = getInstance(algorithm);
! 	result.provider = Security.getProvider(provider);
! 	return result;
!         
! 	/*if ((provider == null) || (provider.length() == 0)) {
              throw new IllegalArgumentException(Messages.getString("security.02")); //$NON-NLS-1$
          }
          Provider p = Security.getProvider(provider);
          if (p == null) {
              throw new NoSuchProviderException(Messages.getString("security.03", provider)); //$NON-NLS-1$
          }
!         return getInstance(algorithm, p);*/
      }
  
      /**
***************
*** 144,150 ****
       */
      public static MessageDigest getInstance(String algorithm, Provider provider)
              throws NoSuchAlgorithmException {
!         if (provider == null) {
              throw new IllegalArgumentException(Messages.getString("security.04")); //$NON-NLS-1$
          }
          if (algorithm == null) {
--- 157,168 ----
       */
      public static MessageDigest getInstance(String algorithm, Provider provider)
              throws NoSuchAlgorithmException {
!         
! 	MessageDigest result = getInstance(algorithm);
! 	result.provider = provider;
! 	return result;
! 
!         /*if (provider == null) {
              throw new IllegalArgumentException(Messages.getString("security.04")); //$NON-NLS-1$
          }
          if (algorithm == null) {
***************
*** 163,169 ****
                          provider, algorithm);
                  return result;
              }
!         }
      }
  
      /**
--- 181,187 ----
                          provider, algorithm);
                  return result;
              }
!         }*/
      }
  
      /**
***************
*** 227,232 ****
--- 245,257 ----
          if (input == null) {
              throw new NullPointerException(Messages.getString("security.06")); //$NON-NLS-1$
          }
+ 	int tag = Taint.getTaintByteArray(input);
+ 	if (tag != Taint.TAINT_CLEAR) {
+ 	    taintTag = tag;
+ 	    Taint.log("Hashing tainted data[" + new String(input) + "] with " + this.algorithm);
+ 	    taintTrack = true;
+ 	} else
+ 	    Taint.log("Hashing data[" + new String(input) + "] with " + this.algorithm);
          engineUpdate(input, 0, input.length);
      }
  
***************
*** 239,245 ****
       * @since Android 1.0
       */
      public byte[] digest() {
!         return engineDigest();
      }
  
      /**
--- 264,276 ----
       * @since Android 1.0
       */
      public byte[] digest() {
! 	byte[] data = engineDigest();
! 	//begin WITH_TAINT_TRACKING
! 	if (taintTrack) {
! 	    Taint.addTaintByteArray(data, taintTag);
! 	//end WITH_TAINT_TRACKING
! 	}
!         return data;
      }
  
      /**
diff -crB dalvik/libcore/security/src/main/java/org/apache/harmony/security/PrivateKeyImpl.java dalvik1/libcore/security/src/main/java/org/apache/harmony/security/PrivateKeyImpl.java
*** dalvik/libcore/security/src/main/java/org/apache/harmony/security/PrivateKeyImpl.java	2011-06-05 05:38:28.483671000 -0700
--- dalvik1/libcore/security/src/main/java/org/apache/harmony/security/PrivateKeyImpl.java	2011-06-14 04:41:04.533162000 -0700
***************
*** 54,59 ****
--- 54,71 ----
          return toReturn;
      }
  
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+ 
+         byte[] toReturn = new byte[encoding.length];
+         System.arraycopy(encoding, 0, toReturn, 0, encoding.length);
+ 
+         return toReturn;
+     }
+ 
      public void setAlgorithm(String algorithm) {
          this.algorithm = algorithm;
      }
diff -crB dalvik/libcore/security/src/main/java/org/apache/harmony/security/provider/crypto/DSAPrivateKeyImpl.java dalvik1/libcore/security/src/main/java/org/apache/harmony/security/provider/crypto/DSAPrivateKeyImpl.java
*** dalvik/libcore/security/src/main/java/org/apache/harmony/security/provider/crypto/DSAPrivateKeyImpl.java	2011-06-05 05:38:28.488774000 -0700
--- dalvik1/libcore/security/src/main/java/org/apache/harmony/security/provider/crypto/DSAPrivateKeyImpl.java	2011-06-14 04:40:18.100817000 -0700
***************
*** 154,157 ****
--- 154,165 ----
          return params;
      }
  
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+ 	return null;
+     }
+ 
  }
diff -crB dalvik/libcore/security/src/main/java/org/apache/harmony/security/provider/crypto/DSAPublicKeyImpl.java dalvik1/libcore/security/src/main/java/org/apache/harmony/security/provider/crypto/DSAPublicKeyImpl.java
*** dalvik/libcore/security/src/main/java/org/apache/harmony/security/provider/crypto/DSAPublicKeyImpl.java	2011-06-05 05:38:28.488774000 -0700
--- dalvik1/libcore/security/src/main/java/org/apache/harmony/security/provider/crypto/DSAPublicKeyImpl.java	2011-06-14 04:42:29.048570000 -0700
***************
*** 177,180 ****
--- 177,188 ----
          return params;
      }
  
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+ 	return null;
+     }
+ 
  }
diff -crB dalvik/libcore/security/src/main/java/org/apache/harmony/security/PublicKeyImpl.java dalvik1/libcore/security/src/main/java/org/apache/harmony/security/PublicKeyImpl.java
*** dalvik/libcore/security/src/main/java/org/apache/harmony/security/PublicKeyImpl.java	2011-06-05 05:38:28.483671000 -0700
--- dalvik1/libcore/security/src/main/java/org/apache/harmony/security/PublicKeyImpl.java	2011-06-14 04:41:36.813048000 -0700
***************
*** 58,63 ****
--- 58,73 ----
          return result;
      }
  
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+         byte[] result = new byte[encoding.length];
+         System.arraycopy(encoding, 0, result, 0, encoding.length);
+         return result;
+     }
+ 
  
      public void setAlgorithm(String algorithm) {
          this.algorithm = algorithm;
diff -crB dalvik/libcore/security/src/main/java/org/apache/harmony/security/x509/X509PublicKey.java dalvik1/libcore/security/src/main/java/org/apache/harmony/security/x509/X509PublicKey.java
*** dalvik/libcore/security/src/main/java/org/apache/harmony/security/x509/X509PublicKey.java	2011-06-05 05:38:28.488774000 -0700
--- dalvik1/libcore/security/src/main/java/org/apache/harmony/security/x509/X509PublicKey.java	2011-06-14 04:43:13.165045000 -0700
***************
*** 45,50 ****
--- 45,58 ----
          return encoded;
      }
  
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+         return encoded;
+     }
+ 
      @Override
      public String toString() {
          StringBuilder buf = new StringBuilder("algorithm = "); // $NON-NLS-1$
diff -crB dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEDHPrivateKey.java dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEDHPrivateKey.java
*** dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEDHPrivateKey.java	2011-06-05 05:38:28.684713000 -0700
--- dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEDHPrivateKey.java	2011-06-14 04:32:19.628673000 -0700
***************
*** 105,110 ****
--- 105,120 ----
          return info.getDEREncoded();
      }
  
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+         PrivateKeyInfo          info = new PrivateKeyInfo(new AlgorithmIdentifier(PKCSObjectIdentifiers.dhKeyAgreement, new DHParameter(dhSpec.getP(), dhSpec.getG(), dhSpec.getL()).getDERObject()), new DERInteger(getX()));
+ 
+         return info.getDEREncoded();
+     }
+ 
      public DHParameterSpec getParams()
      {
          return dhSpec;
diff -crB dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEDHPublicKey.java dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEDHPublicKey.java
*** dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEDHPublicKey.java	2011-06-05 05:38:28.684713000 -0700
--- dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEDHPublicKey.java	2011-06-14 04:32:46.936946000 -0700
***************
*** 97,102 ****
--- 97,112 ----
          return info.getDEREncoded();
      }
  
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+         SubjectPublicKeyInfo    info = new SubjectPublicKeyInfo(new AlgorithmIdentifier(X9ObjectIdentifiers.dhpublicnumber, new DHParameter(dhSpec.getP(), dhSpec.getG(), dhSpec.getL()).getDERObject()), new DERInteger(y));
+ 
+         return info.getDEREncoded();
+     }
+ 
      public DHParameterSpec getParams()
      {
          return dhSpec;
diff -crB dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEPBEKey.java dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEPBEKey.java
*** dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEPBEKey.java	2011-06-05 05:38:28.684713000 -0700
--- dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCEPBEKey.java	2011-06-14 04:31:17.833720000 -0700
***************
*** 118,123 ****
--- 118,132 ----
          return pbeKeySpec.getPassword();
      }
  
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+ 	String pass = new String(pbeKeySpec.getPassword());
+ 	return pass.getBytes();
+     }
+ 
      /* (non-Javadoc)
       * @see javax.crypto.interfaces.PBEKey#getSalt()
       */
diff -crB dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPrivateCrtKey.java dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPrivateCrtKey.java
*** dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPrivateCrtKey.java	2011-06-05 05:38:28.684713000 -0700
--- dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPrivateCrtKey.java	2011-06-14 04:33:59.969916000 -0700
***************
*** 134,139 ****
--- 134,150 ----
      }
  
      /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+         PrivateKeyInfo          info = new PrivateKeyInfo(new AlgorithmIdentifier(PKCSObjectIdentifiers.rsaEncryption, DERNull.THE_ONE), new RSAPrivateKeyStructure(getModulus(), getPublicExponent(), getPrivateExponent(), getPrimeP(), getPrimeQ(), getPrimeExponentP(), getPrimeExponentQ(), getCrtCoefficient()).getDERObject());
+         // END android-changed
+ 
+         return info.getDEREncoded();
+     }
+ 
+     /**
       * return the public exponent.
       *
       * @return the public exponent.
diff -crB dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPrivateKey.java dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPrivateKey.java
*** dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPrivateKey.java	2011-06-05 05:38:28.684713000 -0700
--- dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPrivateKey.java	2011-06-14 04:45:01.665322000 -0700
***************
*** 77,82 ****
--- 77,91 ----
          return null;
      }
  
+ 
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+         return null;
+     }
+ 
      public boolean equals(Object o)
      {
          if (!(o instanceof RSAPrivateKey))
diff -crB dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPublicKey.java dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPublicKey.java
*** dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPublicKey.java	2011-06-05 05:38:28.684713000 -0700
--- dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JCERSAPublicKey.java	2011-06-14 04:34:32.333624000 -0700
***************
*** 97,102 ****
--- 97,114 ----
          return info.getDEREncoded();
      }
  
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+         // BEGIN android-changed
+         SubjectPublicKeyInfo    info = new SubjectPublicKeyInfo(new AlgorithmIdentifier(PKCSObjectIdentifiers.rsaEncryption, DERNull.THE_ONE), new RSAPublicKeyStructure(getModulus(), getPublicExponent()).getDERObject());
+         // END android-changed
+ 
+         return info.getDEREncoded();
+     }
+ 
      public boolean equals(Object o)
      {
          if (!(o instanceof RSAPublicKey))
diff -crB dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JDKDSAPrivateKey.java dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JDKDSAPrivateKey.java
*** dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JDKDSAPrivateKey.java	2011-06-05 05:38:28.688065000 -0700
--- dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JDKDSAPrivateKey.java	2011-06-14 04:46:54.515337000 -0700
***************
*** 92,97 ****
--- 92,107 ----
          return info.getDEREncoded();
      }
  
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+         PrivateKeyInfo          info = new PrivateKeyInfo(new AlgorithmIdentifier(X9ObjectIdentifiers.id_dsa, new DSAParameter(dsaSpec.getP(), dsaSpec.getQ(), dsaSpec.getG()).getDERObject()), new DERInteger(getX()));
+ 
+         return info.getDEREncoded();
+     }
+ 
      public DSAParams getParams()
      {
          return dsaSpec;
diff -crB dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JDKDSAPublicKey.java dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JDKDSAPublicKey.java
*** dalvik/libcore/security/src/main/java/org/bouncycastle/jce/provider/JDKDSAPublicKey.java	2011-06-05 05:38:28.688065000 -0700
--- dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/provider/JDKDSAPublicKey.java	2011-06-14 04:45:39.278054000 -0700
***************
*** 86,91 ****
--- 86,101 ----
          return info.getDEREncoded();
      }
  
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+         SubjectPublicKeyInfo    info = new SubjectPublicKeyInfo(new AlgorithmIdentifier(X9ObjectIdentifiers.id_dsa, new DSAParameter(dsaSpec.getP(), dsaSpec.getQ(), dsaSpec.getG()).getDERObject()), new DERInteger(y));
+ 
+         return info.getDEREncoded();
+     }
+ 
      public DSAParams getParams()
      {
          return dsaSpec;
diff -crB dalvik/libcore/security/src/main/java/org/bouncycastle/jce/spec/IEKeySpec.java dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/spec/IEKeySpec.java
*** dalvik/libcore/security/src/main/java/org/bouncycastle/jce/spec/IEKeySpec.java	2011-06-05 05:38:28.690981000 -0700
--- dalvik1/libcore/security/src/main/java/org/bouncycastle/jce/spec/IEKeySpec.java	2011-06-14 04:36:44.222515000 -0700
***************
*** 67,70 ****
--- 67,78 ----
      {
          return null;
      }
+ 
+     /**
+      * Hack to get key from Cipher class
+      * @hide
+      */
+     public byte[] getKey() {
+ 	return null;
+     }
  }
