diff -crB dalvik/libcore/luni/src/main/java/java/io/FileDescriptor.java dalvik1/libcore/luni/src/main/java/java/io/FileDescriptor.java
*** dalvik/libcore/luni/src/main/java/java/io/FileDescriptor.java	2011-06-05 05:50:50.082570000 -0700
--- dalvik1/libcore/luni/src/main/java/java/io/FileDescriptor.java	2011-06-13 04:46:43.720408000 -0700
***************
*** 79,84 ****
--- 79,85 ----
       */
      public FileDescriptor() {
          super();
+ 	this.id++;
      }
  
      /**
***************
*** 115,126 ****
--- 116,145 ----
      public boolean hasName = false;
  
      /**
+      * Hack for printing out port number
+      * @hide
+      */
+     public int port = 0;
+ 
+     /**
+      * Hack for keeping track of descriptors
+      * @hide
+      */
+     public static int id = 0;
+ 
+     /**
       * hack for printing out IP address
       * @hide
       */
      public String name = null;
  
      /**
+      * hack for buffering read data
+      * @hide
+      */
+     public String readBuffer = "";
+ 
+     /**
       * hack for setting file taint
       * @hide
       */
diff -crB dalvik/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSNetworkSystem.java dalvik1/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSNetworkSystem.java
*** dalvik/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSNetworkSystem.java	2011-06-05 05:50:50.082570000 -0700
--- dalvik1/libcore/luni/src/main/java/org/apache/harmony/luni/platform/OSNetworkSystem.java	2011-06-13 04:51:16.212172000 -0700
***************
*** 116,123 ****
      public int connect(FileDescriptor fd, int trafficClass,
              InetAddress inetAddress, int port) throws IOException{
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostAddress();
  	if (addr != null) {
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
--- 116,124 ----
      public int connect(FileDescriptor fd, int trafficClass,
              InetAddress inetAddress, int port) throws IOException{
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostName();
  	if (addr != null) {
+             fd.port = port;
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
***************
*** 131,138 ****
      public void connectDatagram(FileDescriptor fd, int port,
              int trafficClass, InetAddress inetAddress) throws SocketException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostAddress();
  	if (addr != null) {
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
--- 132,140 ----
      public void connectDatagram(FileDescriptor fd, int port,
              int trafficClass, InetAddress inetAddress) throws SocketException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostName();
  	if (addr != null) {
+ 	    fd.port = port;
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
***************
*** 147,154 ****
              int aport, int timeout, int trafficClass, InetAddress inetAddress)
              throws IOException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostAddress();
  	if (addr != null) {
  	    aFD.hasName = true;
  	    aFD.name = addr;
  	}
--- 149,157 ----
              int aport, int timeout, int trafficClass, InetAddress inetAddress)
              throws IOException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostName();
  	if (addr != null) {
+ 	    //fd.port = port;
  	    aFD.hasName = true;
  	    aFD.name = addr;
  	}
***************
*** 167,174 ****
              int trafficClass, InetAddress inetAddress, int port, int step,
              byte[] context) throws IOException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostAddress();
  	if (addr != null) {
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
--- 170,178 ----
              int trafficClass, InetAddress inetAddress, int port, int step,
              byte[] context) throws IOException {
  	// begin WITH_TAINT_TRACKING
! 	String addr = inetAddress.getHostName();
  	if (addr != null) {
+ 	    fd.port = port;
  	    fd.hasName = true;
  	    fd.name = addr;
  	}
***************
*** 408,414 ****
       */
      public int read(FileDescriptor fd, byte[] data, int offset, int count,
              int timeout) throws IOException {
!         return readSocketImpl(fd, data, offset, count, timeout);
      }
  
      static native int readSocketImpl(FileDescriptor aFD, byte[] data,
--- 412,423 ----
       */
      public int read(FileDescriptor fd, byte[] data, int offset, int count,
              int timeout) throws IOException {
! 	int bytesRead = readSocketImpl(fd, data, offset, count, timeout);
! 	if (bytesRead > 0) {
! 	    String dstr = new String(data, offset, count);
! 	    Taint.log("OSNetworkSystem.read(). Response=["+dstr+"] from " + fd.name + ":" + fd.port);
! 	}
!         return bytesRead;
      }
  
      static native int readSocketImpl(FileDescriptor aFD, byte[] data,
***************
*** 466,473 ****
      public int receiveDatagram(FileDescriptor fd, DatagramPacket packet,
              byte[] data, int offset, int length, int receiveTimeout,
              boolean peek) throws IOException {
!         return receiveDatagramImpl(fd, packet, data, offset, length,
                  receiveTimeout, peek);
      }
  
      static native int receiveDatagramImpl(FileDescriptor aFD,
--- 475,488 ----
      public int receiveDatagram(FileDescriptor fd, DatagramPacket packet,
              byte[] data, int offset, int length, int receiveTimeout,
              boolean peek) throws IOException {
! 
! 	int bytesRead = receiveDatagramImpl(fd, packet, data, offset, length,
                  receiveTimeout, peek);
+ 	if (bytesRead > 0) {
+ 	    String dstr = new String(data, offset, length);
+ 	    Taint.log("OSNetworkSystem.receiveDatagram(). Response=["+dstr+"] from " + fd.name + ":" + fd.port);
+ 	}
+         return bytesRead;
      }
  
      static native int receiveDatagramImpl(FileDescriptor aFD,
***************
*** 507,513 ****
       */
      public int receiveStream(FileDescriptor aFD, byte[] data,
              int offset, int count, int timeout) throws IOException {
!         return receiveStreamImpl(aFD, data, offset, count, timeout);
      }
  
      static native int receiveStreamImpl(FileDescriptor aFD, byte[] data,
--- 522,535 ----
       */
      public int receiveStream(FileDescriptor aFD, byte[] data,
              int offset, int count, int timeout) throws IOException {
! 
! 	int bytesRead = receiveStreamImpl(aFD, data, offset, count, timeout);
! 	if (bytesRead > 0) {
! 	    String dstr = new String(data, offset, count);
! 	    aFD.readBuffer += dstr;
!             Taint.log("OSNetworkSystem.receiveStream(). Response=["+dstr+"] from " + aFD.name + ":" + aFD.port + " ID: " + aFD.id);
!         }
!         return bytesRead;
      }
  
      static native int receiveStreamImpl(FileDescriptor aFD, byte[] data,
***************
*** 528,540 ****
      public int sendStream(FileDescriptor fd, byte[] data, int offset, int count)
              throws IOException {
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
- 	    String dstr = new String(data);
- 	    String addr = (fd.hasName) ? fd.name : "unknown";
  	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.sendStream("+addr+") received data with tag " + tstr + " data=["+dstr+"]");
! 	}
  	// end WITH_TAINT_TRACKING
          return sendStreamImpl(fd, data, offset, count);
      }
--- 550,564 ----
      public int sendStream(FileDescriptor fd, byte[] data, int offset, int count)
              throws IOException {
  	// begin WITH_TAINT_TRACKING
+ 	String dstr = new String(data, offset, count);
+ 	String addr = (fd.hasName) ? fd.name : "unknown";
+ 	int port = (fd.hasName) ? fd.port : 0;
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
  	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.sendStream("+addr+":"+port+") sending data with tag " + tstr + " data=["+dstr+"]");
! 	} else
!             Taint.log("OSNetworkSystem.sendStream("+addr+":"+port+") sending data=["+dstr+"]");
  	// end WITH_TAINT_TRACKING
          return sendStreamImpl(fd, data, offset, count);
      }
***************
*** 663,674 ****
              int offset, int length, boolean bindToDevice) throws IOException {
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
! 	if (tag != Taint.TAINT_CLEAR) {
! 	    String dstr = new String(data);
! 	    String addr = (fd.hasName) ? fd.name : "unknown";
! 	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.sendConnectedDatagram("+addr+") received data with tag " + tstr + " data=["+dstr+"]");
! 	}
  	// end WITH_TAINT_TRACKING
          return sendConnectedDatagramImpl(fd, data, offset, length, bindToDevice);
      }
--- 687,700 ----
              int offset, int length, boolean bindToDevice) throws IOException {
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
!         String dstr = new String(data);
!         String addr = (fd.hasName) ? fd.name : "unknown";
! 	int port = (fd.hasName) ? fd.port : 0;
! 	String tstr = "0x" + Integer.toHexString(tag);
! 	if (tag != Taint.TAINT_CLEAR)
! 	    Taint.log("OSNetworkSystem.sendStream("+addr+":"+port+") sending data with tag " + tstr + " data=["+dstr+"]");
! 	else
! 	    Taint.log("OSNetworkSystem.sendStream("+addr+":"+port+") data=["+dstr+"]");
  	// end WITH_TAINT_TRACKING
          return sendConnectedDatagramImpl(fd, data, offset, length, bindToDevice);
      }
***************
*** 722,731 ****
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
! 	    String dstr = new String(data);
! 	    String addr = (fd.hasName) ? fd.name : "unknown";
! 	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.sendDatagram("+addr+") received data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return sendDatagramImpl(fd, data, offset, length, port, bindToDevice,
--- 748,758 ----
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
! 		String dstr = new String(data);
! 		String addr = (fd.hasName) ? fd.name : "unknown";
! 		int por = (fd.hasName) ? fd.port : 0;
! 		String tstr = "0x" + Integer.toHexString(tag);
! 		Taint.log("OSNetworkSystem.sendStream("+addr+":"+por+") sending data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return sendDatagramImpl(fd, data, offset, length, port, bindToDevice,
***************
*** 741,750 ****
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
! 	    String dstr = new String(data);
! 	    String addr = (fd.hasName) ? fd.name : "unknown";
! 	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.sendDatagram2("+addr+") received data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return sendDatagramImpl2(fd, data, offset, length, port, inetAddress);
--- 768,778 ----
  	// begin WITH_TAINT_TRACKING
  	int tag = Taint.getTaintByteArray(data);
  	if (tag != Taint.TAINT_CLEAR) {
! 		String dstr = new String(data);
! 		String addr = (fd.hasName) ? fd.name : "unknown";
! 		int por = (fd.hasName) ? fd.port : 0;
! 		String tstr = "0x" + Integer.toHexString(tag);
! 		Taint.log("OSNetworkSystem.sendStream("+addr+":"+por+") sending data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return sendDatagramImpl2(fd, data, offset, length, port, inetAddress);
***************
*** 770,783 ****
              int trafficClass, InetAddress inetAddress) throws IOException;
  
      public void sendUrgentData(FileDescriptor fd, byte value) {
- 	// begin WITH_TAINT_TRACKING
- 	int tag = Taint.getTaintByte(value);
- 	String addr = (fd.hasName) ? fd.name : "unknown";
- 	if (tag != Taint.TAINT_CLEAR) {
- 	    String tstr = "0x" + Integer.toHexString(tag);
- 	    Taint.log("OSNetworkSystem.sendUrgentData("+addr+") received data with tag " + tstr + " value=["+value+"]");
- 	}
- 	// end WITH_TAINT_TRACKING
          sendUrgentDataImpl(fd, value);
      }
  
--- 798,803 ----
***************
*** 891,897 ****
  	    String dstr = new String(data);
  	    String addr = (fd.hasName) ? fd.name : "unknown";
  	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.write("+addr+") received data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return writeSocketImpl(fd, data, offset, count);
--- 911,917 ----
  	    String dstr = new String(data);
  	    String addr = (fd.hasName) ? fd.name : "unknown";
  	    String tstr = "0x" + Integer.toHexString(tag);
! 	    Taint.log("OSNetworkSystem.write("+addr+") writing data with tag " + tstr + " data=["+dstr+"]");
  	}
  	// end WITH_TAINT_TRACKING
          return writeSocketImpl(fd, data, offset, count);
diff -crB dalvik/libcore/security/src/main/java/java/security/MessageDigest.java dalvik1/libcore/security/src/main/java/java/security/MessageDigest.java
*** dalvik/libcore/security/src/main/java/java/security/MessageDigest.java	2011-06-05 05:38:28.464405000 -0700
--- dalvik1/libcore/security/src/main/java/java/security/MessageDigest.java	2011-06-13 06:52:54.884148000 -0700
***************
*** 23,29 ****
  package java.security;
  
  import java.nio.ByteBuffer;
! 
  import org.apache.harmony.security.fortress.Engine;
  import org.apache.harmony.security.internal.nls.Messages;
  
--- 23,29 ----
  package java.security;
  
  import java.nio.ByteBuffer;
! import dalvik.system.Taint;
  import org.apache.harmony.security.fortress.Engine;
  import org.apache.harmony.security.internal.nls.Messages;
  
***************
*** 49,54 ****
--- 49,57 ----
      // The algorithm.
      private String algorithm;
  
+     // Taint track hash
+     private boolean taintTrack;
+ 
      /**
       * Constructs a new instance of {@code MessageDigest} with the name of
       * the algorithm to use.
***************
*** 59,64 ****
--- 62,68 ----
       */
      protected MessageDigest(String algorithm) {
          this.algorithm = algorithm;
+         taintTrack = false;
      }
  
      /**
***************
*** 116,129 ****
       */
      public static MessageDigest getInstance(String algorithm, String provider)
              throws NoSuchAlgorithmException, NoSuchProviderException {
!         if ((provider == null) || (provider.length() == 0)) {
              throw new IllegalArgumentException(Messages.getString("security.02")); //$NON-NLS-1$
          }
          Provider p = Security.getProvider(provider);
          if (p == null) {
              throw new NoSuchProviderException(Messages.getString("security.03", provider)); //$NON-NLS-1$
          }
!         return getInstance(algorithm, p);
      }
  
      /**
--- 120,138 ----
       */
      public static MessageDigest getInstance(String algorithm, String provider)
              throws NoSuchAlgorithmException, NoSuchProviderException {
! 
! 	MessageDigest result = getInstance(algorithm);
! 	result.provider = Security.getProvider(provider);
! 	return result;
!         
! 	/*if ((provider == null) || (provider.length() == 0)) {
              throw new IllegalArgumentException(Messages.getString("security.02")); //$NON-NLS-1$
          }
          Provider p = Security.getProvider(provider);
          if (p == null) {
              throw new NoSuchProviderException(Messages.getString("security.03", provider)); //$NON-NLS-1$
          }
!         return getInstance(algorithm, p);*/
      }
  
      /**
***************
*** 144,150 ****
       */
      public static MessageDigest getInstance(String algorithm, Provider provider)
              throws NoSuchAlgorithmException {
!         if (provider == null) {
              throw new IllegalArgumentException(Messages.getString("security.04")); //$NON-NLS-1$
          }
          if (algorithm == null) {
--- 153,164 ----
       */
      public static MessageDigest getInstance(String algorithm, Provider provider)
              throws NoSuchAlgorithmException {
!         
! 	MessageDigest result = getInstance(algorithm);
! 	result.provider = provider;
! 	return result;
! 
!         /*if (provider == null) {
              throw new IllegalArgumentException(Messages.getString("security.04")); //$NON-NLS-1$
          }
          if (algorithm == null) {
***************
*** 163,169 ****
                          provider, algorithm);
                  return result;
              }
!         }
      }
  
      /**
--- 177,183 ----
                          provider, algorithm);
                  return result;
              }
!         }*/
      }
  
      /**
***************
*** 227,232 ****
--- 241,252 ----
          if (input == null) {
              throw new NullPointerException(Messages.getString("security.06")); //$NON-NLS-1$
          }
+ 	int tag = Taint.getTaintByteArray(input);
+ 	if (tag != Taint.TAINT_CLEAR) {
+ 	    Taint.log("Hashing tainted data[" + new String(input) + "] with " + this.algorithm);
+ 	    taintTrack = true;
+ 	} else
+ 	    Taint.log("Hashing data[" + new String(input) + "] with " + this.algorithm);
          engineUpdate(input, 0, input.length);
      }
  
***************
*** 239,245 ****
       * @since Android 1.0
       */
      public byte[] digest() {
!         return engineDigest();
      }
  
      /**
--- 259,280 ----
       * @since Android 1.0
       */
      public byte[] digest() {
! 	byte[] data = engineDigest();
! 	//begin WITH_TAINT_TRACKING
! 	if (taintTrack) {
! 	    Taint.addTaintString(new String(data), Taint.TAINT_IMEI);
!             StringBuffer hexString = new StringBuffer();
!             for (int i = 0; i < data.length; i++) {
!                 String h = Integer.toHexString(0xFF & data[i]);
!                 while (h.length() < 2)
!                     h = "0" + h;
!                 hexString.append(h);
!             }
!             String hash = hexString.toString();
! 	    Taint.log("Taint tracking hash[" + hash + "]");
! 	//end WITH_TAINT_TRACKING
! 	}
!         return data;
      }
  
      /**
